keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,579 @@
+2. Reduce load on API server in such cases.
+3. Reduce memory usage
+
+##### Informers and Caches
+To reduce the need to list all Jobs and CronJobs frequently to reconcile, we propose to replace it with an Informers and WorkQueue based architecture. We would be sharing the [same informer cache](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L450) as the Job controller uses. 
+
+We will follow a controller structure similar to existing workload controllers and as [outlined in the guideline](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure).
+```golang
+
+// 
+type CronJobController struct {
+	kubeClient 			clientset.Interface
+	queue	   			workqueue.RateLimitingInterface
+	recorder   			record.EventRecorder
+	
+	jobControl 			jobControlInterface
+	cronJobControl  		cronJobControlInterface
+	
+	jobLister  			batchv1listers.JobLister
+	cronJobLister 			CronJobLister
+	
+	jobListerSynced 		cache.InformerSynced
+	cronJobListerSynced		cache.InformerSynced
+}
+
+// CronJobControllerConfiguration contains controller config
+type CronJobControllerConfiguration struct {
+	// concurrentCronJobSyncs is the number of cronjob objects that are
+	// allowed to sync concurrently. Larger number = more responsive
+	// but more CPU (and network) load.
+	ConcurrentCronJobSyncs int32
+}
+
+
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, cjc.jobListerSynced, cjc.cronJobListerSynced) {
+		return
+	}
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+func (cjc *CronJobController) func worker() {
+    for cjc.processNextWorkItem() {}
+}
+
+func (cjc *CronJobController) func processNextWorkItem() {
+	key, quit := cjc.queue.Get()
+	if quit {
+		return false
+	}
+	defer cjc.queue.Done(key)
+	if err := cjc.sync(key.(string)); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		cjc.queue.AddRateLimited(key)
+	} else {
+		cjc.queue.Forget(key)
+	}
+	return true
+}
+
+func (cjc *CronJobController)  sync(cronJobKey) {
+	childrenJobs := cjc.jobLister.GetJobsForCronJob(cronJobKey)
+	// reuse/refactor the existing syncOne function
+}
+
+
+```
+
+
+##### Multiple workers
+We also propose to have multiple workers controller by a flag similar to [statefulset controller](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/apps.go#L65). The default would be set to 5 similar to [statefulset](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/config/v1alpha1/defaults.go#L34)
+
+```
+// Controller code
+
+func (cjc *CronJobController)  Run() {
+	...
+    for i := 0; i < workers; i++ {
+        go wait.Until(cjc.worker, time.Second, stopCh)
+    }
+...
+}
+
+```
```

Comment:
this code was already present above - let's avoid duplication

Reply:
removed.